didman/didman.go

```diff
@@ -66,25 +73,49 @@ func (d *didman) Name() string {
 
 func (d *didman) AddEndpoint(id did.DID, serviceType string, u url.URL) error {
 	logging.Log().Debugf("Adding endpoint (did: %s, type: %s, url: %s)", id.String(), serviceType, u.String())
-	doc, meta, err := d.docResolver.Resolve(id, nil)
-	if err != nil {
-		return err
+	err := d.addService(id, serviceType, u.String(), nil)
+	if err == nil {
+		logging.Log().Infof("Endpoint added (did: %s, type: %s, url: %s)", id.String(), serviceType, u.String())
 	}
+	return err
+}
 
-	// check for duplicate service type
-	for _, s := range doc.Service {
-		if s.Type == serviceType {
-			return ErrDuplicateService
+func (d *didman) AddCompoundService(id did.DID, serviceType string, references map[string]ssi.URI) error {
+	logging.Log().Debugf("Adding compound service (did: %s, type: %s, references: %v)", id.String(), serviceType, references)
+
+	// Resolve and validate all service references: they must all resolve to a single endpoint URL
+	// Cache resolved DID documents because most of the time a compound service will refer the same DID document in all service references.
+	documents := make(map[string]*did.Document)
+	var err error
+	for _, serviceRef := range references {
+		referencedDIDStr := serviceRef
+		referencedDIDStr.RawQuery = ""
+		referencedDIDStr.Fragment = ""
+		var document *did.Document
+		if document = documents[referencedDIDStr.String()]; document == nil {
+			referencedDID, err := did.ParseDID(referencedDIDStr.String())
+			if err != nil {
+				return err
+			}
+			document, _, err = d.docResolver.Resolve(*referencedDID, nil)
+			if err != nil {
+				return err
+			}
+			documents[referencedDIDStr.String()] = document
+		}
+		queriedServiceType := serviceRef.Query().Get("type")
+		if len(queriedServiceType) == 0 {
+			return ErrInvalidServiceQuery
+		}
+		_, _, err := document.ResolveEndpointURL(queriedServiceType)
+		if err != nil {
+			return errors2.WithMessage(ErrReferencedServiceNotAnEndpoint, err.Error())
```

Comment:
We have a mix of `github.com/pkg/errors` and `errors` inour code. `github.com/pkg/errors` had its uses in older versions of Go. I think it would be best to stick to `errors` and `fmt.Errorf()`. That way `errors.Is` can be used.

In the case of this error, a distinction needs to be made between a technical error or a ErrNotFound/ErrDupl.

Reply:
Fixed it. About the distinction I don't agree: it's up to the caller to interpret the particular error and call it again after fixing the input. And since the caller will (most probably) be a REST client it doesn't really add value to specialize these errors any further?

Reply:
I see it's not possible to distinguish between different errors and that no technical error can occur.